File: backend/Deprecated/sources.js
import { Router } from 'express';
import multer from 'multer';
import csv from 'csvtojson';

// Accept requireAdmin middleware injected from server.js
export default function (pool, requireAdmin) {
    const router = Router();

    // File upload limits and MIME type validation
    const upload = multer({
        storage: multer.memoryStorage(),
        limits: {
            fileSize: 50 * 1024 * 1024, // 50 MB max
        },
        fileFilter: (req, file, cb) => {
            if (!file.originalname.match(/\.csv$/i)) {
                return cb(new Error('Only .csv files are allowed.'));
            }
            cb(null, true);
        }
    });

    // Ensure the metadata table exists and all context columns are present (idempotent)
    pool.query(`
        CREATE TABLE IF NOT EXISTS bfi.sources_meta (
            id SERIAL PRIMARY KEY,
            name VARCHAR(255) NOT NULL,
            table_name VARCHAR(255) UNIQUE NOT NULL,
            status VARCHAR(50) DEFAULT 'Ready',
            size BIGINT,
            num_rows INT,
            columns JSONB,
            uploaded_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
        );
    `)
    .then(() => pool.query(`
        ALTER TABLE bfi.sources_meta
            ADD COLUMN IF NOT EXISTS project_name  VARCHAR(255),
            ADD COLUMN IF NOT EXISTS description   TEXT,
            ADD COLUMN IF NOT EXISTS data_domain   VARCHAR(100),
            ADD COLUMN IF NOT EXISTS coverage_start DATE,
            ADD COLUMN IF NOT EXISTS coverage_end   DATE,
            ADD COLUMN IF NOT EXISTS ongoing        BOOLEAN DEFAULT FALSE,
            ADD COLUMN IF NOT EXISTS agency_response VARCHAR(50);
    `))
    .catch(err => console.error('Failed to initialise sources_meta table:', err));

    // Upload a CSV — requires admin role (requireAdmin checks JWT on the backend)
    router.post('/', requireAdmin, upload.single('file'), async (req, res) => {
        if (!req.file) {
            return res.status(400).json({ message: 'No file uploaded.' });
        }

        // Parse the uploaded CSV in memory
        let rows;
        try {
            rows = await csv({ trim: true, ignoreEmpty: true })
                .fromString(req.file.buffer.toString('utf8'));
        } catch (parseErr) {
            return res.status(400).json({ message: `Failed to parse CSV: ${parseErr.message}` });
        }

        if (rows.length === 0) {
            return res.status(400).json({ message: 'CSV contained no data rows.' });
        }

        // C1: Extract tenant from JWT and validate — never interpolate unsanitised schema names into SQL.
        const SAFE_SCHEMA = /^[a-z][a-z0-9_]{0,62}$/;
        const tenant = SAFE_SCHEMA.test(req.user?.tenant) ? req.user.tenant : 'bfi';

        // M-2: enforce 60-char table name limit to prevent PostgreSQL silent truncation
        //      and identifier collisions (PG max is 63 chars).
        const rawName = req.file.originalname.replace(/[^a-zA-Z0-9]/g, '_').toLowerCase();
        const tableName = rawName.slice(0, 60);
        const rawColumns = Object.keys(rows[0]);

        // B4: Sanitize CSV column names — they are untrusted user input and get
        // interpolated into SQL identifiers. Even with quoting, a column name containing
        // a double-quote character can break out of a quoted identifier.
        const sanitizeColumnName = (name, idx) => {
            const cleaned = String(name)
                .replace(/[^a-zA-Z0-9_ ]/g, '')  // strip all non-safe chars (including quotes)
                .trim()
                .replace(/\s+/g, '_')             // normalise spaces to underscores
                .slice(0, 63);                    // PostgreSQL max identifier length
            return cleaned.length > 0 ? cleaned : `col_${idx}`;
        };
        const columns = rawColumns.map(sanitizeColumnName);

        // Use a transaction so the entire upload is atomic —
        // if anything fails, the old table is not dropped and no partial state is left.
        const client = await pool.connect();
        try {
            await client.query('BEGIN');

            // M-1: Acquire a transaction-level advisory lock keyed on the table name.
            // This serialises concurrent uploads of the same filename so the
            // DROP TABLE / CREATE TABLE sequence is never interleaved.
            await client.query('SELECT pg_advisory_xact_lock(hashtext($1))', [`bfi.${tableName}`]);

            // 1. Drop existing table if it exists
            await client.query(`DROP TABLE IF EXISTS ${tenant}."${tableName}";`);

            // 2. Build CREATE TABLE query dynamically based on the CSV columns
            const createCols = columns.map(col => `"${col}" TEXT`).join(', ');
            await client.query(`CREATE TABLE ${tenant}."${tableName}" (id SERIAL PRIMARY KEY, ${createCols});`);

            // 3. Bulk insert rows using a single multi-row INSERT per batch (5000 rows/batch)
            // This replaces the old row-by-row approach which caused N round-trips to the DB.
            const BATCH_SIZE = 5000;
            for (let offset = 0; offset < rows.length; offset += BATCH_SIZE) {
                const batch = rows.slice(offset, offset + BATCH_SIZE);
                const params = [];
                const valuePlaceholders = [];
                let paramIndex = 1;

                for (const row of batch) {
                    const rowPlaceholders = columns.map(() => `$${paramIndex++}`);
                    valuePlaceholders.push(`(${rowPlaceholders.join(', ')})`);
                    columns.forEach(col => params.push(row[col] ?? null));
                }

                const colList = columns.map(c => `"${c}"`).join(', ');
                await client.query(
                    `INSERT INTO ${tenant}."${tableName}" (${colList}) VALUES ${valuePlaceholders.join(', ')};`,
                    params
                );
            }

            // 4. Save metadata to sources_meta — inside the same transaction
            const metaResult = await client.query(`
                INSERT INTO bfi.sources_meta (name, table_name, size, num_rows, columns)
                VALUES ($1, $2, $3, $4, $5)
                ON CONFLICT (table_name) DO UPDATE
                    SET name = EXCLUDED.name, size = EXCLUDED.size,
                        num_rows = EXCLUDED.num_rows, columns = EXCLUDED.columns,
                        uploaded_at = CURRENT_TIMESTAMP
                RETURNING id;
            `, [
                req.file.originalname,
                tableName,
                req.file.size,
                rows.length,
                JSON.stringify(columns)
            ]);

            await client.query('COMMIT');

            res.status(201).json({
                message: 'SQL Table created successfully!',
                _id: metaResult.rows[0].id,
                name: req.file.originalname,
                status: 'Ready',
                schema: tenant,
                table: tableName,
                num_rows: rows.length,
                size: req.file.size,
                columns
            });

        } catch (error) {
            await client.query('ROLLBACK');
            console.error('Upload Error:', error);
            res.status(500).json({ message: error.message });
        } finally {
            client.release();
        }
    });

    // List all uploaded sources (metadata) — any authenticated user can view
    router.get('/', async (req, res) => {
        const SAFE_SCHEMA = /^[a-z][a-z0-9_]{0,62}$/;
        const tenant = SAFE_SCHEMA.test(req.user?.tenant) ? req.user.tenant : 'bfi';
        try {
            const result = await pool.query(
                `SELECT * FROM ${tenant}.sources_meta ORDER BY uploaded_at DESC`
            );
            res.json(result.rows);
        } catch (error) {
            res.status(500).json({ message: error.message });
        }
    });

    // Delete a source — requires admin role
    router.delete('/:id', requireAdmin, async (req, res) => {
        const SAFE_SCHEMA = /^[a-z][a-z0-9_]{0,62}$/;
        const tenant = SAFE_SCHEMA.test(req.user?.tenant) ? req.user.tenant : 'bfi';
        try {
            const id = parseInt(req.params.id, 10);
            if (isNaN(id)) return res.status(400).json({ message: 'Invalid source id.' });

            const metaResult = await pool.query(`SELECT table_name FROM ${tenant}.sources_meta WHERE id = $1`, [id]);
            if (metaResult.rows.length === 0) {
                return res.status(404).json({ message: 'Source not found.' });
            }
            const tableName = metaResult.rows[0].table_name;

            // Drop table and metadata in a transaction
            const client = await pool.connect();
            try {
                await client.query('BEGIN');
                await client.query(`DROP TABLE IF EXISTS ${tenant}."${tableName}";`);
                await client.query(`DELETE FROM ${tenant}.sources_meta WHERE id = $1`, [id]);
                await client.query('COMMIT');
                res.json({ deleted: true });
            } catch (err) {
                await client.query('ROLLBACK');
                throw err;
            } finally {
                client.release();
            }
        } catch (error) {
            res.status(500).json({ message: error.message });
        }
    });

    // Update submission context metadata (project name, description, dates, AI consent).
    // Called immediately after a successful upload when the Submission Context modal is submitted.
    // Requires admin role (same as upload).
    router.patch('/:id/context', requireAdmin, async (req, res) => {
        const SAFE_SCHEMA = /^[a-z][a-z0-9_]{0,62}$/;
        const tenant = SAFE_SCHEMA.test(req.user?.tenant) ? req.user.tenant : 'bfi';
        try {
            const id = parseInt(req.params.id, 10);
            if (isNaN(id)) return res.status(400).json({ message: 'Invalid source id.' });

            const {
                projectName,
                description,
                dataDomain,
                coverageStart,
                coverageEnd,
                ongoing,
                agencyResponse,
            } = req.body;

            // Only update the fields that are safe to store; ignore unknown keys.
            const result = await pool.query(`
                UPDATE ${tenant}.sources_meta
                SET
                    project_name   = $1,
                    description    = $2,
                    data_domain    = $3,
                    coverage_start = $4,
                    coverage_end   = $5,
                    ongoing        = $6,
                    agency_response = $7
                WHERE id = $8
                RETURNING id, name, project_name, description, data_domain,
                          coverage_start, coverage_end, ongoing, agency_response;
            `, [
                projectName   || null,
                description   || null,
                dataDomain    || null,
                coverageStart || null,
                coverageEnd   || null,
                ongoing       === true,
                agencyResponse || null,
                id,
            ]);

            if (result.rows.length === 0) {
                return res.status(404).json({ message: 'Source not found.' });
            }

            res.json(result.rows[0]);
        } catch (error) {
            console.error('Context update error:', error);
            res.status(500).json({ message: error.message });
        }
    });

    return router;
}

File: backend/sources.py
"""
Data Hub source uploads (port of sources.js).

Exposes a blueprint factory. The `require_admin` decorator is injected from
app.py so RBAC stays defined in one place. The authenticated user lives on
`flask.g.user` (set by app.py's auth middleware).
"""

import csv
import io
import json
import re

from flask import Blueprint, g, jsonify, request
from psycopg2.extras import execute_values

import db

SAFE_SCHEMA = re.compile(r"^[a-z][a-z0-9_]{0,62}$")
MAX_FILE_SIZE = 50 * 1024 * 1024  # 50 MB
BATCH_SIZE = 5000


def _tenant():
    """Extract and validate the tenant schema from the JWT claims."""
    t = (g.user or {}).get("tenant")
    return t if t and SAFE_SCHEMA.match(t) else "bfi"


def _sanitize_column_name(name, idx):
    cleaned = re.sub(r"[^a-zA-Z0-9_ ]", "", str(name)).strip()
    cleaned = re.sub(r"\s+", "_", cleaned)[:63]
    return cleaned if cleaned else "col_{}".format(idx)


def init_sources_meta():
    """Ensure the metadata table exists with all context/RBAC columns (idempotent)."""
    try:
        db.query(
            """
            CREATE TABLE IF NOT EXISTS bfi.sources_meta (
                id SERIAL PRIMARY KEY,
                name VARCHAR(255) NOT NULL,
                table_name VARCHAR(255) UNIQUE NOT NULL,
                status VARCHAR(50) DEFAULT 'Ready',
                size BIGINT,
                num_rows INT,
                columns JSONB,
                uploaded_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            );
            """
        )
        db.query(
            """
            ALTER TABLE bfi.sources_meta
                ADD COLUMN IF NOT EXISTS user_id         INTEGER REFERENCES users(id) ON DELETE SET NULL,
                ADD COLUMN IF NOT EXISTS visibility      VARCHAR(20) DEFAULT 'private',
                ADD COLUMN IF NOT EXISTS project_name    VARCHAR(255),
                ADD COLUMN IF NOT EXISTS description     TEXT,
                ADD COLUMN IF NOT EXISTS data_domain     VARCHAR(100),
                ADD COLUMN IF NOT EXISTS coverage_start  DATE,
                ADD COLUMN IF NOT EXISTS coverage_end    DATE,
                ADD COLUMN IF NOT EXISTS ongoing         BOOLEAN DEFAULT FALSE,
                ADD COLUMN IF NOT EXISTS agency_response VARCHAR(50);
            """
        )
        db.query(
            """
            CREATE INDEX IF NOT EXISTS idx_sources_meta_user       ON bfi.sources_meta(user_id);
            CREATE INDEX IF NOT EXISTS idx_sources_meta_visibility ON bfi.sources_meta(visibility);
            """
        )
    except Exception as e:
        print("Failed to initialise sources_meta table:", e)


def create_sources_blueprint(require_admin, require_editor=None, require_analyzer=None, require_viewer=None):
    bp = Blueprint("sources", __name__)
    # Fall back to require_admin if narrower role guards are not provided.
    require_editor = require_editor or require_admin

    # Upload a CSV — editors and admins
    @bp.route("", methods=["POST"])
    @bp.route("/", methods=["POST"])
    @require_editor
    def upload():
        file = request.files.get("file")
        if not file:
            return jsonify({"message": "No file uploaded."}), 400

        if not re.search(r"\.csv$", file.filename or "", re.IGNORECASE):
            return jsonify({"message": "Only .csv files are allowed."}), 400

        raw = file.read()
        if len(raw) > MAX_FILE_SIZE:
            return jsonify({"message": "File too large (max 50 MB)."}), 400

        # Parse the uploaded CSV in memory
        try:
            text = raw.decode("utf-8")
            reader = csv.DictReader(io.StringIO(text))
            raw_columns = reader.fieldnames or []
            rows = [
                {k: (v.strip() if isinstance(v, str) else v) for k, v in r.items()}
                for r in reader
            ]
        except Exception as parse_err:
            return jsonify({"message": "Failed to parse CSV: {}".format(parse_err)}), 400

        if not rows:
            return jsonify({"message": "CSV contained no data rows."}), 400

        tenant = _tenant()

        # M-2: enforce 60-char table name limit (PG max identifier is 63).
        raw_name = re.sub(r"[^a-zA-Z0-9]", "_", file.filename).lower()
        table_name = raw_name[:60]

        # B4: sanitize untrusted CSV column names before interpolating into SQL.
        columns = [_sanitize_column_name(c, i) for i, c in enumerate(raw_columns)]

        size = len(raw)
        # Visibility: admin uploads are shared with all users; editor uploads are private.
        visibility = "shared" if (g.user or {}).get("role") == "admin" else "private"

        try:
            with db.transaction() as cur:
                # M-1: advisory lock serialises concurrent uploads of the same filename.
                cur.execute("SELECT pg_advisory_xact_lock(hashtext(%s))", ("bfi.{}".format(table_name),))

                # 1. Drop existing table if it exists
                cur.execute('DROP TABLE IF EXISTS {}."{}";'.format(tenant, table_name))

                # 2. Build CREATE TABLE dynamically from the CSV columns
                create_cols = ", ".join('"{}" TEXT'.format(c) for c in columns)
                cur.execute(
                    'CREATE TABLE {}."{}" (id SERIAL PRIMARY KEY, {});'.format(
                        tenant, table_name, create_cols
                    )
                )

                # 3. Bulk insert rows in batches via execute_values
                col_list = ", ".join('"{}"'.format(c) for c in columns)
                insert_sql = 'INSERT INTO {}."{}" ({}) VALUES %s'.format(
                    tenant, table_name, col_list
                )
                for offset in range(0, len(rows), BATCH_SIZE):
                    batch = rows[offset:offset + BATCH_SIZE]
                    values = [
                        tuple(row.get(c) for c in columns) for row in batch
                    ]
                    execute_values(cur, insert_sql, values)

                # 4. Save metadata in the same transaction - track owner and visibility
                cur.execute(
                    """
                    INSERT INTO bfi.sources_meta (name, table_name, size, num_rows, columns, user_id, visibility)
                    VALUES (%s, %s, %s, %s, %s, %s, %s)
                    ON CONFLICT (table_name) DO UPDATE
                        SET name        = EXCLUDED.name,
                            size        = EXCLUDED.size,
                            num_rows    = EXCLUDED.num_rows,
                            columns     = EXCLUDED.columns,
                            user_id     = EXCLUDED.user_id,
                            visibility  = EXCLUDED.visibility,
                            uploaded_at = CURRENT_TIMESTAMP
                    RETURNING id;
                    """,
                    (file.filename, table_name, size, len(rows), json.dumps(columns), (g.user or {}).get("id"), visibility),
                )
                new_id = cur.fetchone()["id"]

            return jsonify({
                "message": "SQL Table created successfully!",
                "_id": new_id,
                "name": file.filename,
                "status": "Ready",
                "schema": tenant,
                "table": table_name,
                "num_rows": len(rows),
                "size": size,
                "columns": columns,
            }), 201
        except Exception as error:
            print("Upload Error:", error)
            return jsonify({"message": str(error)}), 500

    # List all uploaded sources — any authenticated user
    @bp.route("", methods=["GET"])
    @bp.route("/", methods=["GET"])
    def list_sources():
        tenant = _tenant()
        try:
            user_role = (g.user or {}).get("role")
            user_id = (g.user or {}).get("id")
            if user_role == "admin":
                rows = db.query(
                    "SELECT * FROM {}.sources_meta ORDER BY uploaded_at DESC".format(tenant)
                )
            else:
                rows = db.query(
                    "SELECT * FROM {}.sources_meta WHERE user_id = %s ORDER BY uploaded_at DESC".format(tenant),
                    (user_id,),
                )
            return jsonify(rows)
        except Exception as error:
            return jsonify({"message": str(error)}), 500

    # Delete a source — admin only
    @bp.route("/<id>", methods=["DELETE"])
    @require_editor
    def delete_source(id):
        tenant = _tenant()
        try:
            source_id = int(id)
        except (TypeError, ValueError):
            return jsonify({"message": "Invalid source id."}), 400
        try:
            meta = db.query(
                "SELECT table_name, user_id FROM {}.sources_meta WHERE id = %s".format(tenant),
                (source_id,),
            )
            if not meta:
                return jsonify({"message": "Source not found."}), 404

            # Ownership check: editors can only delete their own sources
            user_role = (g.user or {}).get("role")
            user_id = (g.user or {}).get("id")
            if user_role != "admin" and meta[0]["user_id"] != user_id:
                return jsonify({"error": "You can only delete your own sources."}), 403
            
            table_name = meta[0]["table_name"]

            with db.transaction() as cur:
                cur.execute('DROP TABLE IF EXISTS {}."{}";'.format(tenant, table_name))
                cur.execute(
                    "DELETE FROM {}.sources_meta WHERE id = %s".format(tenant),
                    (source_id,),
                )
            return jsonify({"deleted": True})
        except Exception as error:
            return jsonify({"message": str(error)}), 500

    # Update submission context metadata — admin only
    @bp.route("/<id>/context", methods=["PATCH"])
    @require_editor
    def update_context(id):
        tenant = _tenant()
        try:
            source_id = int(id)
        except (TypeError, ValueError):
            return jsonify({"message": "Invalid source id."}), 400
        
        # Editors can only update their own uploads; admins can update any.
        user_role = (g.user or {}).get("role")
        user_id = (g.user or {}).get("id")
        if user_role != "admin":
            owner_check = db.query(
                "SELECT user_id FROM {}.sources_meta WHERE id = %s".format(tenant),
                (source_id,),
            )
            if not owner_check:
                return jsonify({"message": "Source not found."}), 404
            if owner_check[0]["user_id"] != user_id:
                return jsonify({"message": "You can only update your own sources."}), 403

        body = request.get_json(silent=True) or {}
        try:
            rows = db.query(
                """
                UPDATE {}.sources_meta
                SET
                    project_name    = %s,
                    description     = %s,
                    data_domain     = %s,
                    coverage_start  = %s,
                    coverage_end    = %s,
                    ongoing         = %s,
                    agency_response = %s
                WHERE id = %s
                RETURNING id, name, project_name, description, data_domain,
                          coverage_start, coverage_end, ongoing, agency_response;
                """.format(tenant),
                (
                    body.get("projectName") or None,
                    body.get("description") or None,
                    body.get("dataDomain") or None,
                    body.get("coverageStart") or None,
                    body.get("coverageEnd") or None,
                    body.get("ongoing") is True,
                    body.get("agencyResponse") or None,
                    source_id,
                ),
            )
            if not rows:
                return jsonify({"message": "Source not found."}), 404
            return jsonify(rows[0])
        except Exception as error:
            print("Context update error:", error)
            return jsonify({"message": str(error)}), 500

    return bp
